open_lm/file_utils.py

```diff
@@ -111,15 +111,39 @@ def get_shards_for_epoch(num_samples, epoch, path):
             chunk_count_list.append(real_chunk_count)
             real_chunk_count = 0
     
-    return shard_list[epoch % len(shard_list)], chunk_count_list[epoch % len(chunk_count_list)]
-
-def get_string_for_epoch(num_samples, epoch, path):
-    shard_list, num_samples = get_shards_for_epoch(num_samples, epoch, path)
-    shard_root = '/'.join(path.split('/')[:-1]) + '/shard_'
-    shard_string = shard_root + '{' + shard_list[0] + '..' + shard_list[-1] + '}.tar'
-    if path.startswith('s3'):
-        shard_string = f'pipe:aws s3 cp {shard_string} -'
-    return shard_string, num_samples
+    return shard_list[chunk % len(shard_list)], chunk_count_list[chunk % len(chunk_count_list)]
+
+
+def enough_shards(shard_lists, min_shards_needed):
+    for sl in shard_lists:
+        if len(sl) < min_shards_needed:
+            return False
+    return True
+
+def get_string_for_epoch(num_samples, starting_chunk, paths, weights, min_shards_needed):
+    samples_per_source = [weights[i] * num_samples / sum(weights) for i in range(len(weights))]
+    
+    shard_strings_per_source = []
+    next_chunk = starting_chunk
+    shard_list_per_source = [[] for _ in range(len(paths))]
+    num_samples_per_source = [0 for _ in range(len(paths))]
+    while not enough_shards(shard_list_per_source, min_shards_needed):
```

Comment:
is it possible this will infinite loop? maybe add some error handling in that case? what happens in the case where epochs*num_train_tokens > number of tokens in the dataset?

Reply:
Good point - some error handling is needed here.

Reply:
Added some additional error handling, and found a corner case.
If you run out of shards and have to loop back, it is kind of problematic with the current string notation.

I updated the notation to list all shards e.g. ```pipe aws s3 cp s3://shards/shard_{00000, 00002, 00004}.tar -```, which I think has the desired effect of allowing arbitrary shards (not just continuous ones).